tests/lib/prepare.sh

```diff
@@ -518,3 +529,266 @@ prepare_all_snap() {
+            grep -v "^root:" "$UNPACKD/etc/$f" > /mnt/system-data/root/test-etc/"$f"
+            # append this systems root user so that linode can connect
+            grep "^root:" /etc/"$f" >> /mnt/system-data/root/test-etc/"$f"
+
+            # make sure the group is as expected
+            chgrp --reference "$UNPACKD/etc/$f" /mnt/system-data/root/test-etc/"$f"
+            # now bind mount read-only those passwd files on boot
+            cat >/mnt/system-data/etc/systemd/system/etc-"$f".mount <<EOF
+[Unit]
+Description=Mount root/test-etc/$f over system etc/$f
+Before=ssh.service
+
+[Mount]
+What=/root/test-etc/$f
+Where=/etc/$f
+Type=none
+Options=bind,ro
+
+[Install]
+WantedBy=multi-user.target
+EOF
+            ln -s /etc/systemd/system/etc-"$f".mount /mnt/system-data/etc/systemd/system/multi-user.target.wants/etc-"$f".mount
+
+            # create /var/lib/extrausers/$f
+            # append ubuntu, test user for the testing
+            grep "^test:" /etc/$f >> /mnt/system-data/var/lib/extrausers/"$f"
+            grep "^ubuntu:" /etc/$f >> /mnt/system-data/var/lib/extrausers/"$f"
+            # check test was copied
+            MATCH "^test:" </mnt/system-data/var/lib/extrausers/"$f"
+            MATCH "^ubuntu:" </mnt/system-data/var/lib/extrausers/"$f"
+        done
+
+        # ensure spread -reuse works in the core image as well
+        if [ -e /.spread.yaml ]; then
+            cp -av /.spread.yaml /mnt/system-data
+        fi
+
+        # using symbolic names requires test:test have the same ids
+        # inside and outside which is a pain (see 12345 above), but
+        # using the ids directly is the wrong kind of fragile
+        chown --verbose test:test /mnt/user-data/test
+
+        # we do what sync-dirs is normally doing on boot, but because
+        # we have subdirs/files in /etc/systemd/system (created below)
+        # the writeable-path sync-boot won't work
+        mkdir -p /mnt/system-data/etc/systemd
+        (cd /tmp ; unsquashfs -v  /var/lib/snapd/snaps/core18_*.snap etc/systemd/system)
+        cp -avr /tmp/squashfs-root/etc/systemd/system /mnt/system-data/etc/systemd/
+
+
+        umount /mnt
+        kpartx -d  "$IMAGE_HOME/$IMAGE"
+
+        # the reflash magic
+        # FIXME: ideally in initrd, but this is good enough for now
+        cat > "$IMAGE_HOME/reflash.sh" << EOF
+#!/bin/sh -ex
+mount -t tmpfs none /tmp
+cp /bin/busybox /tmp
+cp $IMAGE_HOME/$IMAGE /tmp
+sync
+# blow away everything
+/tmp/busybox dd if=/tmp/$IMAGE of=/dev/sda bs=4M
+# and reboot
+/tmp/busybox sync
+/tmp/busybox echo b > /proc/sysrq-trigger
+EOF
+        chmod +x "$IMAGE_HOME/reflash.sh"
+
+        # extract ROOT from /proc/cmdline
+        ROOT=$(sed -e 's/^.*root=//' -e 's/ .*$//' /proc/cmdline)
+        cat >/boot/grub/grub.cfg <<EOF
+set default=0
+set timeout=2
+menuentry 'flash-all-snaps' {
+linux /vmlinuz root=$ROOT ro init=$IMAGE_HOME/reflash.sh console=ttyS0
+initrd /initrd.img
+}
+EOF
+}
+
+prepare_core18() {
```

Comment:
it is pretty similat to prepare_all_snaps, perhaps in the future we could unify both, what do you think?

Reply:
Absolutely! Thanks for this suggestion. I had a bit of an uneasy feeling when adding so much duplicated code. It was mostly a shortcut until GCE tests work, I have now unified the two codepaths and only the (relatively small) differences are different.